olive/resource_path.py

```diff
@@ -0,0 +1,379 @@
+
+        # path to save the resource to
+        if name:
+            new_path_name = Path(name).with_suffix(self.path.suffix).name
+        else:
+            new_path_name = self.path.name
+        new_path = dir_path / new_path_name
+        _overwrite_helper(new_path, overwrite)
+
+        # is the resource a file or a folder
+        is_file = Path(self.path).is_file()
+        # copy the resource to the new path
+        if is_file:
+            shutil.copy(self.path, new_path)
+        else:
+            shutil.copytree(self.path, new_path)
+
+        return str(new_path)
+
+    @validator("path")
+    def path_must_exist(cls, v: Union[Path, str]) -> Union[Path, str]:
+        """Validate that the path exists."""
+        path = Path(v)
+        if not path.exists():
+            raise FileNotFoundError(f"{path} does not exist.")
+        return path
+
+    @validator("path")
+    def resolve_path(cls, v: Union[Path, str]) -> Union[Path, str]:
+        """Resolve the path."""
+        return Path(v).resolve()
+
+
+class LocalFileConfig(LocalResourceConfig):
+    """Config for a local file."""
+
+    _type = ResourceType.LocalFile
+
+    @validator("path")
+    def path_must_be_file(cls, v: Union[Path, str]) -> Union[Path, str]:
+        """Validate that the path is a file."""
+        path = Path(v)
+        if not path.is_file():
+            raise ValueError(f"Path {path} is not a file.")
+        return path
+
+
+class LocalFolderConfig(LocalResourceConfig):
+    """Config for a local folder."""
+
+    _type = ResourceType.LocalFolder
+
+    @validator("path")
+    def path_must_be_folder(cls, v: Union[Path, str]) -> Union[Path, str]:
+        """Validate that the path is a folder."""
+        path = Path(v)
+        if not path.is_dir():
+            raise ValueError(f"Path {path} is not a folder.")
+        return path
+
+
+class StringNameConfig(ResourceConfig):
+    """Config for a string name."""
+
+    _type = ResourceType.StringName
+    name: str = Field(..., description="Name of the resource.")
+
+    def get_path(self) -> str:
+        return self.name
+
+    def save_to_dir(self, dir_path: Union[Path, str], name: str = None, overwrite: bool = False) -> str:
+        logger.debug("Resource is a string name. No need to save to directory.")
+        return self.name
+
+
+def _get_azureml_resource_prefix(workspace_config: Dict[str, str]) -> str:
+    return (
+        f"azureml://subscriptions/{workspace_config['subscription_id']}"
+        f"/resourcegroups/{workspace_config['resource_group']}"
+        f"/workspaces/{workspace_config['workspace_name']}"
+    )
+
+
+class AzureMLModelConfig(ResourceConfig):
+    """Config for an AzureML model."""
+
+    _type = ResourceType.AzureMLModel
+    azureml_client: AzureMLClientConfig = Field(..., description="AzureML client config.")
+    name: str = Field(..., description="Name of the model.")
+    version: Union[int, str] = Field(..., description="Version of the model.")
```

Comment:
Double check: will the version be required for AML model? What if None?

Reply:
Yes version is required. I think that is safer since if we default to something like `latest`, we cannot guarantee caching. The `latest` version might change.